modules/message-lane/src/lib.rs

```diff
@@ -0,0 +1,274 @@
+	}
+}
+
+// Pallets use events to inform users when important changes are made.
+// https://substrate.dev/docs/en/knowledgebase/runtime/events
+decl_event!(
+	pub enum Event {
+		/// Message has been accepted and is waiting to be delivered.
+		MessageAccepted(LaneId, MessageNonce),
+		/// Messages in the inclusive range have been delievered to the bridged chain.
+		MessagesDelivered(LaneId, MessageNonce, MessageNonce),
+		/// Messages in the inclusive range have been processed by the bridged chain.
+		MessagesProcessed(LaneId, MessageNonce, MessageNonce),
+	}
+);
+
+decl_module! {
+	pub struct Module<T: Trait<I>, I: Instance = DefaultInstance> for enum Call where origin: T::Origin {
+		/// Deposit one of this module's events by using the default implementation.
+		fn deposit_event() = default;
+
+		/// Send message over lane.
+		#[weight = 0] // TODO: update me (https://github.com/paritytech/parity-bridges-common/issues/78)
+		pub fn send_message(
+			origin,
+			lane_id: LaneId,
+			payload: T::Payload,
+		) {
+			let _ = ensure_signed(origin)?;
+			let mut lane = outbound_lane::<T, I>(lane_id);
+			let nonce = lane.send_message(payload);
+			lane.prune_messages(T::MaxMessagesToPruneAtOnce::get());
+
+			Self::deposit_event(Event::MessageAccepted(lane_id, nonce));
+		}
+	}
+}
+
+impl<T: Trait<I>, I: Instance> Module<T, I> {
+	// =========================================================================================
+	// === Exposed mutables ====================================================================
+	// =========================================================================================
+
+	/// Receive new TRUSTED lane messages.
+	///
+	/// Trusted here means that the function itself doesn't check whether message has actually
+	/// been sent through the other end of the channel. We only check that we are receiving
+	/// and processing messages in order here.
+	///
+	/// Messages vector is required to be sorted by nonce within each lane. Otherise messages
+	/// will be rejected.
+	pub fn receive_messages(messages: Vec<Message<T::Payload>>) -> MessageNonce {
+		let mut correct_messages = 0;
+		let mut processor = T::OnMessageReceived::default();
+		for message in messages {
+			let mut lane = inbound_lane::<T, I>(message.key.lane_id);
+			if lane.receive_message(message.key.nonce, message.payload, &mut processor) {
+				correct_messages += 1;
+			}
+		}
+
+		correct_messages
+	}
+
+	/// Process stored lane messages.
+	///
+	/// Stops processing either when all messages are processed, or when processor returns
+	/// MessageResult::NotProcessed.
+	pub fn process_lane_messages(lane_id: &LaneId, processor: &mut impl OnMessageReceived<T::Payload>) {
+		inbound_lane::<T, I>(*lane_id).process_messages(processor);
+	}
+
+	/// Receive TRUSTED proof of message receival.
+	///
+	/// Trusted here means that the function itself doesn't check whether the bridged chain has
+	/// actually received these messages.
```

Comment:
Would it be the job of the relay to make sure that the messages are in fact in storage before making this call? I imagine this function would (or should) be called immediately after `receive_messages`, right?

Reply:
> Would it be the job of the relay to make sure that the messages are in fact in storage before making this call?

Nope. The pallet has zero trust to relay. What relay does is very dumb routine - it checks whether any new messages have appeared in target storage. If there are new messages, it prepares a proof-of-storage (proof of inbound lane state) and submits tx to source node. The pallet will verify proof-of-storage (using your pallet :) ) and finally call this function.

Right now the pallet **API** is unfinished. As I said [here](https://github.com/paritytech/parity-bridges-common/pull/299#issuecomment-678325725), initially I thought that it'll be `BridgePallet` who will call exposed mutables, but actually it should be the opposite. So in next PR I'd add some trait like `trait DummyOrderedChain { fn check_message_proof() {} fn check_inbound_lane_state_proof() {} ... }`. Its implementation would be based on top of your Substrate-spceific `HeaderChain` implementation. And then I'll add more calls to this pallet - i.e. all what's currently called 'exposed mutables' would be callable using transactions.

Again - if you insist, I could add this trait to this PR. But would prefer to do this in follow-up. Changes wouldn't be huge.

> I imagine this function would (or should) be called immediately after receive_messages, right?

It is up to relayer to decide when to call this function. But our implementation would call it as soon as it finds that some messages have been received, right.